docs/dev/text-type.md

```diff
@@ -0,0 +1,45 @@
+## History overview
+
+* [odfe#620](https://github.com/opendistro-for-elasticsearch/sql/pull/620) - added two types for `text` (just text and text with fields), querying and string functions are supported, but searching is not.
+* [odfe#682](https://github.com/opendistro-for-elasticsearch/sql/pull/682) and [odfe#730](https://github.com/opendistro-for-elasticsearch/sql/pull/730) - added support for querying (in 682) and aggregation (730); if `text` has sub-fields, `.keyword` prefix is added to the field name regardless of the sub-field names.
+* [#1314](https://github.com/opensearch-project/sql/pull/1314) and [#1664](https://github.com/opensearch-project/sql/pull/1664) - changed format of storing `text` type; it is one type now, subfield information is stored, but not yet used.
+* Current changes (no PR number yet) - correctly resolve sub-field name if a `text` field has only one subfield. Fixes [#1112](https://github.com/opensearch-project/sql/issues/1112) and [#1038](https://github.com/opensearch-project/sql/issues/1038).
+
+## Further changes
+
+* Support search for text sub-fields ([#1113](https://github.com/opensearch-project/sql/issues/1113)).
+* Support multiple sub-fields for text ([#1887](https://github.com/opensearch-project/sql/issues/1887)).
+* Support non-default date formats for search queries ([#1847](https://github.com/opensearch-project/sql/issues/1847)). Fix for this bug depends on the current changes.
+
+## Problem statement
+
+`:opensearch` module parses index mapping and builds instances of `OpenSearchDataType` (a base class), but ships simplified types (`ExprCoreType` - a enum) to `:core` module, because `:core` uses `ExprCoreType`s to resolve functions.
+
+```mermaid
+sequenceDiagram
+  participant core as :core
+  participant opensearch as :opensearch
+
+  core ->>+ opensearch : resolve types
+  opensearch ->>- core : simplified types
+  note over core: preparing query
+  core ->> opensearch : translate request into datasource DSL
+```
+
+Later, `:core` returns to `:opensearch` the DSL request with types stored. Since types were simplified, all mapping information is lost. Adding new `TEXT` entry to `ExprCoreType` enum is not enough, because `ExprCoreType` is datasource agnostic and can't store any specific mapping info.
+
+## Solution
+
+The solution is to provide to `:core` non simplified types, but full types. Those objects should be fully compatible with `ExprCoreType` and implement all required APIs to allow `:core` to manipulate with built-in functions. Once those type objects are returned back to `:opensearch`, it can get all required information to build the correct search request.
```

Comment:
non simplified types: `enum`
full types: `Objects`
right? 
Can we just say that?

Reply:
simplified types: `enum`
full types: `Objects`

Before full types were converted to enums before passing from `:opensearch` to `:core`. With my changes full types are passed from `:opensearch` to `:core`, and `:core` uses an API call to convert them to a enum value whatever it is needed (to pick proper function signature).